From 812bc03e12984ae44ae815ac1615a335b472a95f Mon Sep 17 00:00:00 2001
From: nift4 <nift4@protonmail.com>
Date: Sat, 15 Jul 2023 15:10:06 +0200
Subject: [PATCH 1/3] Notify AuthController when face providers are ready

This change introduces a callback that AuthController can provide to
FaceService to get notified when all of the face providers
are available.

Please refer to 06f04c041cf65b19cbf736852996760b1284124c ("Notify
AuthController when fingerprint providers are ready") for more
information.

Change-Id: Iaf4523feb7a65f818123b7aa0ab497a8285a5ca0
---
 .../android/hardware/face/FaceManager.java    | 25 ++++++
 ...IFaceAuthenticatorsRegisteredCallback.aidl | 35 ++++++++
 .../android/hardware/face/IFaceService.aidl   |  5 ++
 .../systemui/biometrics/AuthController.java   | 30 ++++++-
 .../biometrics/sensors/face/FaceService.java  | 81 ++++++++++++++++++-
 5 files changed, 170 insertions(+), 6 deletions(-)
 create mode 100644 core/java/android/hardware/face/IFaceAuthenticatorsRegisteredCallback.aidl

diff --git a/core/java/android/hardware/face/FaceManager.java b/core/java/android/hardware/face/FaceManager.java
index 7092e43596ec..de11c7fe8fbd 100644
--- a/core/java/android/hardware/face/FaceManager.java
+++ b/core/java/android/hardware/face/FaceManager.java
@@ -673,6 +673,31 @@ public class FaceManager implements BiometricAuthenticator, BiometricFaceConstan
         return new ArrayList<>();
     }
 
+    /**
+     * Adds a callback that gets called when the service registers all of the face
+     * authenticators (HALs).
+     *
+     * If the face authenticators are already registered when the callback is added, the
+     * callback is invoked immediately.
+     *
+     * The callback is automatically removed after it's invoked.
+     *
+     * @hide
+     */
+    @RequiresPermission(USE_BIOMETRIC_INTERNAL)
+    public void addAuthenticatorsRegisteredCallback(
+            IFaceAuthenticatorsRegisteredCallback callback) {
+        if (mService != null) {
+            try {
+                mService.addAuthenticatorsRegisteredCallback(callback);
+            } catch (RemoteException e) {
+                throw e.rethrowFromSystemServer();
+            }
+        } else {
+            Slog.w(TAG, "addProvidersAvailableCallback(): Service not connected!");
+        }
+    }
+
     /**
      * @hide
      */
diff --git a/core/java/android/hardware/face/IFaceAuthenticatorsRegisteredCallback.aidl b/core/java/android/hardware/face/IFaceAuthenticatorsRegisteredCallback.aidl
new file mode 100644
index 000000000000..d0127df788c6
--- /dev/null
+++ b/core/java/android/hardware/face/IFaceAuthenticatorsRegisteredCallback.aidl
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package android.hardware.face;
+
+import android.hardware.face.FaceSensorPropertiesInternal;
+import java.util.List;
+
+/**
+ * Callback to notify FaceManager that FaceService has registered all of the
+ * face authenticators (HALs).
+ * See {@link android.hardware.fingerprint.IFaceService#registerAuthenticators}.
+ *
+ * @hide
+ */
+oneway interface IFaceAuthenticatorsRegisteredCallback {
+    /**
+     * Notifies FaceManager that all of the face authenticators have been registered.
+     *
+     * @param sensors A consolidated list of sensor properties for all of the authenticators.
+     */
+    void onAllAuthenticatorsRegistered(in List<FaceSensorPropertiesInternal> sensors);
+}
diff --git a/core/java/android/hardware/face/IFaceService.aidl b/core/java/android/hardware/face/IFaceService.aidl
index 989b001ca8bf..16fabd419fae 100644
--- a/core/java/android/hardware/face/IFaceService.aidl
+++ b/core/java/android/hardware/face/IFaceService.aidl
@@ -20,6 +20,7 @@ import android.hardware.biometrics.IBiometricServiceLockoutResetCallback;
 import android.hardware.biometrics.IInvalidationCallback;
 import android.hardware.biometrics.ITestSession;
 import android.hardware.biometrics.ITestSessionCallback;
+import android.hardware.face.IFaceAuthenticatorsRegisteredCallback;
 import android.hardware.face.IFaceServiceReceiver;
 import android.hardware.face.Face;
 import android.hardware.face.FaceSensorPropertiesInternal;
@@ -134,4 +135,8 @@ interface IFaceService {
     // AIDL sensor properties are retrieved directly from the available HALs. If no HIDL HALs exist,
     // hidlSensors must be non-null and empty. See AuthService.java
     void registerAuthenticators(in List<FaceSensorPropertiesInternal> hidlSensors);
+
+    // Adds a callback which gets called when the service registers all of the face
+    // authenticators. The callback is automatically removed after it's invoked.
+    void addAuthenticatorsRegisteredCallback(IFaceAuthenticatorsRegisteredCallback callback);
 }
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/AuthController.java b/packages/SystemUI/src/com/android/systemui/biometrics/AuthController.java
index 3302073af7c6..f14644d2f241 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/AuthController.java
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/AuthController.java
@@ -47,6 +47,7 @@ import android.hardware.biometrics.IBiometricContextListener;
 import android.hardware.biometrics.IBiometricSysuiReceiver;
 import android.hardware.biometrics.PromptInfo;
 import android.hardware.display.DisplayManager;
+import android.hardware.face.IFaceAuthenticatorsRegisteredCallback;
 import android.hardware.face.FaceManager;
 import android.hardware.face.FaceSensorPropertiesInternal;
 import android.hardware.fingerprint.FingerprintManager;
@@ -148,7 +149,7 @@ public class AuthController implements CoreStartable,  CommandQueue.Callbacks,
     @Nullable private IBiometricContextListener mBiometricContextListener;
     @VisibleForTesting IBiometricSysuiReceiver mReceiver;
     @VisibleForTesting @NonNull final BiometricDisplayListener mOrientationListener;
-    @Nullable private final List<FaceSensorPropertiesInternal> mFaceProps;
+    @Nullable private List<FaceSensorPropertiesInternal> mFaceProps;
     @Nullable private List<FingerprintSensorPropertiesInternal> mFpProps;
     @Nullable private List<FingerprintSensorPropertiesInternal> mUdfpsProps;
     @Nullable private List<FingerprintSensorPropertiesInternal> mSidefpsProps;
@@ -195,6 +196,16 @@ public class AuthController implements CoreStartable,  CommandQueue.Callbacks,
                 }
             };
 
+    private final IFaceAuthenticatorsRegisteredCallback
+            mFaceAuthenticatorsRegisteredCallback =
+            new IFaceAuthenticatorsRegisteredCallback.Stub() {
+                @Override
+                public void onAllAuthenticatorsRegistered(
+                        List<FaceSensorPropertiesInternal> sensors) {
+                    mHandler.post(() -> handleAllFaceAuthenticatorsRegistered(sensors));
+                }
+            };
+
     private final BiometricStateListener mBiometricStateListener =
             new BiometricStateListener() {
                 @Override
@@ -330,6 +341,18 @@ public class AuthController implements CoreStartable,  CommandQueue.Callbacks,
         }
     }
 
+    private void handleAllFaceAuthenticatorsRegistered(
+            List<FaceSensorPropertiesInternal> sensors) {
+        mExecution.assertIsMainThread();
+        if (DEBUG) {
+            Log.d(TAG, "handleAllFaceAuthenticatorsRegistered | sensors: " + Arrays.toString(
+                    sensors.toArray()));
+        }
+        mFaceProps = sensors;
+
+        updateSensorLocations();
+    }
+
     private void handleEnrollmentsChanged(int userId, int sensorId, boolean hasEnrollments) {
         mExecution.assertIsMainThread();
         Log.d(TAG, "handleEnrollmentsChanged, userId: " + userId + ", sensorId: " + sensorId
@@ -773,7 +796,6 @@ public class AuthController implements CoreStartable,  CommandQueue.Callbacks,
             }
         });
 
-        mFaceProps = mFaceManager != null ? mFaceManager.getSensorPropertiesInternal() : null;
         int[] faceAuthLocation = context.getResources().getIntArray(
                 com.android.systemui.R.array.config_face_auth_props);
         if (faceAuthLocation == null || faceAuthLocation.length < 2) {
@@ -836,6 +858,10 @@ public class AuthController implements CoreStartable,  CommandQueue.Callbacks,
             mFingerprintManager.addAuthenticatorsRegisteredCallback(
                     mFingerprintAuthenticatorsRegisteredCallback);
         }
+        if (mFaceManager != null) {
+            mFaceManager.addAuthenticatorsRegisteredCallback(
+                    mFaceAuthenticatorsRegisteredCallback);
+        }
 
         mActivityTaskManager.registerTaskStackListener(mTaskStackListener);
         mOrientationListener.enable();
diff --git a/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java b/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java
index 5491379b39ac..de37d07d0a1a 100644
--- a/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java
+++ b/services/core/java/com/android/server/biometrics/sensors/face/FaceService.java
@@ -37,6 +37,7 @@ import android.hardware.biometrics.face.SensorProps;
 import android.hardware.face.Face;
 import android.hardware.face.FaceSensorPropertiesInternal;
 import android.hardware.face.FaceServiceReceiver;
+import android.hardware.face.IFaceAuthenticatorsRegisteredCallback;
 import android.hardware.face.IFaceService;
 import android.hardware.face.IFaceServiceReceiver;
 import android.os.Binder;
@@ -44,6 +45,7 @@ import android.os.Handler;
 import android.os.IBinder;
 import android.os.NativeHandle;
 import android.os.Process;
+import android.os.RemoteCallbackList;
 import android.os.RemoteException;
 import android.os.ServiceManager;
 import android.os.UserHandle;
@@ -52,6 +54,7 @@ import android.util.Slog;
 import android.util.proto.ProtoOutputStream;
 import android.view.Surface;
 
+import com.android.internal.annotations.GuardedBy;
 import com.android.internal.util.DumpUtils;
 import com.android.internal.widget.LockPatternUtils;
 import com.android.server.ServiceThread;
@@ -80,12 +83,20 @@ public class FaceService extends SystemService {
 
     protected static final String TAG = "FaceService";
 
+    private final Object mLock = new Object();
     private final FaceServiceWrapper mServiceWrapper;
     private final LockoutResetDispatcher mLockoutResetDispatcher;
     private final LockPatternUtils mLockPatternUtils;
     @NonNull
     private final List<ServiceProvider> mServiceProviders;
 
+    @GuardedBy("mLock")
+    @NonNull private final RemoteCallbackList<IFaceAuthenticatorsRegisteredCallback>
+            mAuthenticatorsRegisteredCallbacks;
+
+    @GuardedBy("mLock")
+    @NonNull private final List<FaceSensorPropertiesInternal> mSensorProps;
+
     @Nullable
     private ServiceProvider getProviderForSensor(int sensorId) {
         for (ServiceProvider provider : mServiceProviders) {
@@ -123,11 +134,9 @@ public class FaceService extends SystemService {
 
     @NonNull
     private List<FaceSensorPropertiesInternal> getSensorProperties() {
-        final List<FaceSensorPropertiesInternal> properties = new ArrayList<>();
-        for (ServiceProvider provider : mServiceProviders) {
-            properties.addAll(provider.getSensorProperties());
+        synchronized (mLock) {
+            return mSensorProps;
         }
-        return properties;
     }
 
     /**
@@ -616,6 +625,38 @@ public class FaceService extends SystemService {
                     new ClientMonitorCallbackConverter(receiver), opPackageName);
         }
 
+        // Notifies the callbacks that all of the authenticators have been registered and removes the
+        // invoked callbacks from the callback list.
+        private void broadcastAllAuthenticatorsRegistered() {
+            // Make a local copy of the data so it can be used outside of the synchronized block when
+            // making Binder calls.
+            final List<IFaceAuthenticatorsRegisteredCallback> callbacks = new ArrayList<>();
+            final List<FaceSensorPropertiesInternal> props;
+            synchronized (mLock) {
+                if (!mSensorProps.isEmpty()) {
+                    props = new ArrayList<>(mSensorProps);
+                } else {
+                    Slog.e(TAG, "mSensorProps is empty");
+                    return;
+                }
+                final int n = mAuthenticatorsRegisteredCallbacks.beginBroadcast();
+                for (int i = 0; i < n; ++i) {
+                    final IFaceAuthenticatorsRegisteredCallback cb =
+                            mAuthenticatorsRegisteredCallbacks.getBroadcastItem(i);
+                    callbacks.add(cb);
+                    mAuthenticatorsRegisteredCallbacks.unregister(cb);
+                }
+                mAuthenticatorsRegisteredCallbacks.finishBroadcast();
+            }
+            for (IFaceAuthenticatorsRegisteredCallback cb : callbacks) {
+                try {
+                    cb.onAllAuthenticatorsRegistered(props);
+                } catch (RemoteException e) {
+                    Slog.e(TAG, "Remote exception in onAllAuthenticatorsRegistered", e);
+                }
+            }
+        }
+
         private void addHidlProviders(@NonNull List<FaceSensorPropertiesInternal> hidlSensors) {
             for (FaceSensorPropertiesInternal hidlSensor : hidlSensors) {
                 mServiceProviders.add(
@@ -685,8 +726,38 @@ public class FaceService extends SystemService {
                         }
                     }
                 }
+
+                synchronized (mLock) {
+                    for (ServiceProvider provider : mServiceProviders) {
+                        mSensorProps.addAll(provider.getSensorProperties());
+                    }
+                }
+
+                broadcastAllAuthenticatorsRegistered();
             });
         }
+
+        @Override // Binder call
+        public void addAuthenticatorsRegisteredCallback(
+                IFaceAuthenticatorsRegisteredCallback callback) {
+            Utils.checkPermission(getContext(), USE_BIOMETRIC_INTERNAL);
+            if (callback == null) {
+                Slog.e(TAG, "addAuthenticatorsRegisteredCallback, callback is null");
+                return;
+            }
+
+            final boolean registered;
+            final boolean hasSensorProps;
+            synchronized (mLock) {
+                registered = mAuthenticatorsRegisteredCallbacks.register(callback);
+                hasSensorProps = !mSensorProps.isEmpty();
+            }
+            if (registered && hasSensorProps) {
+                broadcastAllAuthenticatorsRegistered();
+            } else if (!registered) {
+                Slog.e(TAG, "addAuthenticatorsRegisteredCallback failed to register callback");
+            }
+        }
     }
 
     public FaceService(Context context) {
@@ -695,6 +766,8 @@ public class FaceService extends SystemService {
         mLockoutResetDispatcher = new LockoutResetDispatcher(context);
         mLockPatternUtils = new LockPatternUtils(context);
         mServiceProviders = new ArrayList<>();
+        mAuthenticatorsRegisteredCallbacks = new RemoteCallbackList<>();
+        mSensorProps = new ArrayList<>();
     }
 
     @Override
-- 
2.29.2

